tests/gem/test_api.py

```diff
@@ -0,0 +1,80 @@
+import asyncio
+import logging
+import socket
+import sys
+import unittest
+from typing import Optional
+
+from siobrultech_protocols.gem.api import DELAY_NEXT_PACKET, GET_SERIAL_NUMBER, GemApi
+from siobrultech_protocols.gem.protocol import PacketProtocol
+from tests.gem.mock_gem import MockGemProtocol
+from tests.gem.packet_test_data import assert_packet, read_packet
+
+logging.basicConfig(
+    stream=sys.stderr,
+    level=logging.DEBUG,
+    format="%(asctime)s [%(name)s](%(levelname)s) %(message)s",
+)
+
+LOG = logging.getLogger(__name__)
+
+
+class TestApi(unittest.IsolatedAsyncioTestCase):
+    async def asyncSetUp(self):
```

Comment:
This method is complicated and took me a little while to read it and comprehend what it is doing.  Can you add some comments to help explain for future us and others to not have to figure out what is going on here when we are debugging a test failure when adding some new feature? :D

Reply:
Most of what makes this complicated (to me anyway) is the stuff to adapt the synchronous callback providing an API instance to the async code that wants to wait for an API to be available.

You solved that for packets with a `Queue`; what do you think about just sending the `GemApi` to the same `Queue` and letting clients multiplex based on the type of the object? Something like:
```python
    message = await queue.get()
    if isinstance(message, Packet):
        # Handle packet
    elif isinstance(message, GemApi):
        # Query serial number and add api to lookup table
```

The test code would be really simple since we control both sides of the connection. The first entry in the queue would always be the `GemApi` so we'd just dequeue that instead of messing with the callbacks and the `Condition`.

Reply:
Went with this; see what you think.

Reply:
I like using a queue, but I think we might want two, and here's my justification (but please disagree if you have a strong opinion here, especially since I don't have a lot of thought into how you are working on this for HA!):
* one queue for `Packet`s and one queue for `GemAPI` means consumers can have two worker tasks with a specific purpose (one for each)
* one queue for both means consumers have to inspect the type of each thing from the queue, and then figure out what to do with it.

With that said, it's easy enough to split if we go with this approach, or merge into one queue if we go with two queues.  Thoughts?

Reply:
I thought about using two. Didn't like having two params of the same type on the constructor (though I guess callers could avoid any possible confusion there using keyword args).

However, I'm not strongly attached to that approach, and anyway your other point is likely to send this in a very different direction.

Reply:
Although as I look at the sans-I/O stuff again, the single queue is effectively an event queue. One of their examples is hyper-h2, which also requires type inspection for handling events ([example](https://python-hyper.org/projects/h2/en/stable/twisted-head-example.html)).

Reply:
I think that's actually cleaner then if we go with single queue.  Wrapping it in some kind of event object with pre-defined types (via an `@enum.unique` `enum.Enum` object) is at least more approachable for end users.  Something like:
```python
from enum import Enum, unique

@unique
class EventType(Enum):
  PACKET_RECEIVED = "PACKET_RECEIVED"
  DEVICE_CONNECTED = "DEVICE_CONNECTED"

class Event:
  def __init__(self, event_type: EventType):
    self._event_type

  @property
  def event_type(self) -> EventType:
    return self._event_type

class PacketReceivedEvent(Event):
  def __init__(self, packet: Packet):
    super().__init__(EventType.PACKET_RECEIVED)
    self._packet = packet

  @property
  def packet(self) -> Packet:
    return self._packet
```

So my take is two clearly defined queues, or one queue wrapped in an event object.  The event object might let us attach more things to pass to consumers too (like maybe the transport or some other data for advanced usage if needed)

> However, I'm not strongly attached to that approach, and anyway your other point is likely to send this in a very different direction.

Honestly, we can merge this regardless, but we probably shouldn't cut a new version until we resolve that.  (a), (c), and (d) can all be done post-landing pretty easily.  I don't want to make more work for you, if I can avoid it.